### PR TITLE
Add --max-memory-usage argument, and associated khmer_args refactoring.

### PR DESCRIPTION
**Pull request by ctb**
This pull request should make little practical difference in behavior, but does several major refactors:

* first, we introduce `-M/--max-memory-usage` which is a single parameter to automatically set
  graph sizes;
* second, as part of this, we change the way that hashtable sizes are calculated so that prime numbers
   *below* the specified number are chosen.  This necessitated many small changes to tests that
   relied on specific behavior and numbers.
* third, I made CountingHash take the same args as Hashbits, and put in the same __new__ hack in
   Python-land to subclass the CPython class _CountingHash;
* fourth, while we were changing so many things anyway, I went ahead and replaced   
  new_counting_hash and new_hashbits with CountingHash and Hashbits.
* fifth, since test_hashbits and test_hashbits_obj now contained basically the same tests, I merged them
  and then removed test_hashbits_obj.

and of course many minor adjustments throughout.

- [x] add `-M/--max-memory-usage` parameter in khmer_args.py (try 2 to address https://github.com/dib-lab/khmer/issues/732#issuecomment-70554583; try 1 is #744)
- [x] start salting in nodegraph and countgraph terminology at Python level, as convenient (ref #872)
- [x] fix up all of the scripts to use `create_nodegraph/create_countgraph` functions from khmer_args.py
- [x] add unit tests for create_nodegraph and create_countgraph
- [x] add tests for -M parameter in scripts/
- [x] change `check_space_for_hashtable` to use 'args' rather than 'min_tablesize'
- [x] then, run test_scripts after renaming the min_tablesize parameter to see what else we need to change
- [x] change new_counting_hash over to CountingHash, and new_hashbits over to Hashbits throughout
- [x] check basic behavior of sandbox/ scripts with recipes and/or manual tests
- [x] update docs
- [x] update whatsnew for 2.0
- [x] adjust -M behavior for special scripts like abundance-dist, which use 1.1x the normal memory.

### CODE REVIEW

**Comment by ctb** [comment]
Fixes #980 in passing.

**Comment by ctb** [comment]
This is one for @mr-c to gaze upon. So far, I am happy with the direction it is taking.

**Comment by ctb** [comment]
I think I will wait on @mr-c to take a look-see before continuing with fixing up the documentation etc. etc.

File: ChangeLog [hunk 1]
```diff
@@ -1,3 +1,23 @@
...
+   * khmer/kfile.py: fixed check_space_for_hashtable to depend on args obj.
+   * tests/test_script_arguments.py: updated tests.
+   * scripts/{abundance-dist-single.py,count-overlap.py,do-partition.py,
+   filter-abund-single.py,find-knots.py,load-into-counting.py,
+   make-initial-stoptags.py,normalize-by-median.py,trim-low-abund.py,
+   sandbox/{collect-reads.py,saturate-by-median.py}: replaced
+   counting hash/hashbits creation with new khmer_args create* functions.
+   * oxli/build_graph.py,sandbox/{sweep-files.py,sweep-reads.py},
+   scripts/{count-overlap.py,do-partition.py,},khmer/khmer_args.py:
+   changed hashtype over to 'nodegraph' and 'bitgraph' in call to
```
Comment (mr-c) [inline ChangeLog]: s/bitgraph/countgraph/

**Comment by ctb** [comment]
ARGH this one is going down the rabbit hole... <sigh> fine, I'll just do it all, then.

**Comment by mr-c** [comment]
Well, all of these inter-related issues provide a lot of value and are a long time coming :-)

**Comment by ctb** [comment]
@mr-c review, please? I have yet to update docs and I have a few tests to add, but otherwise I think the overall PR is close to done.

File: khmer/__init__.py [hunk 2]
```diff
@@ -279,8 +302,8 @@ def __new__(cls, k, starting_size, n_tables):
 class CountingLabelHash(_LabelHash):
 
     def __new__(cls, k, starting_size, n_tables):
-        primes = get_n_primes_above_x(n_tables, starting_size)
```
Comment (mr-c) [inline khmer/__init__.py]: Should we drop `get_n_primes_above_x` ?

File: khmer/kfile.py [hunk 3]
```diff
@@ -111,8 +112,11 @@ def check_space(in_files, force, _testhook_free_space=None):
             sys.exit(1)
 
 
-def check_space_for_hashtable(hash_size, force, _testhook_free_space=None):
+def check_space_for_hashtable(args, hashtype, force,
+                              _testhook_free_space=None):
     """Check we have enough size to write a hash table."""
+    hash_size = khmer_args._calculate_tablesize(args, hashtype)
```
Comment (mr-c) [inline khmer/kfile.py]: We should unprefix `_calculate_tablesize` if it is going to be accessed outside of `khmer_args`

**Comment by mr-c** [comment]
Might as well remove the deprecated factory methods as well.

File: khmer/kfile.py (hunk 3 above)
Comment (ctb) [inline khmer/kfile.py]: On Fri, Jun 19, 2015 at 11:47:24AM -0700, Michael R. Crusoe wrote:
> >      """Check we have enough size to write a hash table."""
> > +    hash_size = khmer_args._calculate_tablesize(args, hashtype)
> 
> We should unprefix `_calculate_tablesize` if it is going to be accessed outside of `khmer_args`

It's only used within khmer, and is not a public function.

File: khmer/__init__.py [hunk 4]
```diff
@@ -56,7 +56,7 @@ def new_hashbits(k, starting_size, n_tables=2):
     starting_size -- lower bound on hashsize to use
     n_tables -- number of hash tables to use (default = 2)
     """
-    primes = get_n_primes_above_x(n_tables, starting_size)
+    primes = get_n_primes_near_x(n_tables, starting_size)
```
Comment (mr-c) [inline khmer/__init__.py]: No test coverage -- delete the function?

File: khmer/__init__.py [hunk 5]
```diff
@@ -70,7 +70,7 @@ def new_counting_hash(k, starting_size, n_tables=2):
     n_tables -- number of hash tables to use (default = 2)
     n_threads  -- number of simultaneous threads to execute (default = 1)
     """
-    primes = get_n_primes_above_x(n_tables, starting_size)
+    primes = get_n_primes_near_x(n_tables, starting_size)
```
Comment (mr-c) [inline khmer/__init__.py]: No test coverage -- delete the function?

File: khmer/__init__.py [hunk 6]
```diff
@@ -237,6 +246,11 @@ def get_n_primes_near_x(number, target):
         if is_prime(i):
             primes.append(i)
         i -= 2
+
+    if len(primes) != number:
+        raise Exception("unable to find %d prime numbers < %d" % (number,
```
Comment (mr-c) [inline khmer/__init__.py]: FYI: No test coverage

File: khmer/khmer_args.py [hunk 7]
```diff
@@ -131,30 +130,71 @@ def __call__(self, parser, namespace, values, option_string=None):
                         action=LoadAction)
 
 
-def report_on_config(args, hashtype='counting'):
+def _calculate_tablesize(args, hashtype):
+    if hashtype not in ('countgraph', 'nodegraph'):
+        raise Exception, "unknown graph type:", hashtype
```
Comment (mr-c) [inline khmer/khmer_args.py]: No test coverage

File: khmer/khmer_args.py [hunk 8]
```diff
@@ -131,30 +130,71 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+
+def report_on_config(args, hashtype='countgraph'):
     """Print out configuration.
 
     Summarize the configuration produced by the command-line arguments
     made available by this module.
     """
     from khmer.utils import print_error
+    if hashtype not in ('countgraph', 'nodegraph'):
+        raise Exception, "unknown graph type:", hashtype
```
Comment (mr-c) [inline khmer/khmer_args.py]: No test coverage

**Comment by mr-c** [comment]
Looking good, so far

**Comment by ctb** [comment]
test coverage issues fixed.

**Comment by ctb** [comment]
test this please, jenkins.

**Comment by bocajnotnef** [comment]
retest this, please

**Comment by ctb** [comment]
Ready for review and merge @mr-c.

**Comment by ctb** [comment]
Note: we will probably need to redo things in doc/user/scripts.rst and choosing-hash-sizes a bit once #1106 lands; please evaluate current doc updates in this light :)

File: doc/whats-new-2.0.rst [hunk 9]
```diff
@@ -3,10 +3,26 @@
...
-and partition subsets) have changed. Files are
-now pre-pended with the string ``OXLI`` to indicate that they are from this
-project.
+Incompatible changes
+====================
+
+New parameter for tablesize/number of table parameters.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There is now a `-M/--max-memory-usage` parameter that sets the number
```
Comment (mr-c) [inline doc/whats-new-2.0.rst]: Should use the style
```
:option:`-M`/:option:`--max-memory-usage`
```
throughout this document.

**Comment by mr-c** [comment]
Great! I'm going to fix the two small nits I found and merge for you.

**Comment by ctb** [comment]
fixed the doc one

On Fri, Jun 26, 2015 at 05:20:09PM -0700, Michael R. Crusoe wrote:
> Great! I'm going to fix the two small nits I found and merge for you.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/khmer/pull/1050#issuecomment-115923701
-- 
C. Titus Brown, ctbrown@ucdavis.edu

**Comment by mr-c** [comment]
*sigh* Thanks :-P

On Fri, Jun 26, 2015 at 5:21 PM C. Titus Brown <notifications@github.com>
wrote:

> fixed the doc one
>
> On Fri, Jun 26, 2015 at 05:20:09PM -0700, Michael R. Crusoe wrote:
> > Great! I'm going to fix the two small nits I found and merge for you.
> >
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/dib-lab/khmer/pull/1050#issuecomment-115923701
> --
> C. Titus Brown, ctbrown@ucdavis.edu
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/1050#issuecomment-115923802>.
>
-- 
Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
The lab for Data Intensive Biology; University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

**Comment by ctb** [comment]
lol

On Fri, Jun 26, 2015 at 05:27:33PM -0700, Michael R. Crusoe wrote:
> *sigh* Thanks :-P
> 
> On Fri, Jun 26, 2015 at 5:21 PM C. Titus Brown <notifications@github.com>
> wrote:
> 
> > fixed the doc one
> >
> > On Fri, Jun 26, 2015 at 05:20:09PM -0700, Michael R. Crusoe wrote:
> > > Great! I'm going to fix the two small nits I found and merge for you.
> > >
> > > ---
> > > Reply to this email directly or view it on GitHub:
> > > https://github.com/dib-lab/khmer/pull/1050#issuecomment-115923701
> > --
> > C. Titus Brown, ctbrown@ucdavis.edu
> >
> > ???
> > Reply to this email directly or view it on GitHub
> > <https://github.com/dib-lab/khmer/pull/1050#issuecomment-115923802>.
> >
> -- 
> Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
> The lab for Data Intensive Biology; University of California, Davis
> https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/khmer/pull/1050#issuecomment-115924190
-- 
C. Titus Brown, ctbrown@ucdavis.edu

**Comment by ctb** [comment]
great, thanks! cc @bocajnotnef

**Pull request merged**